x86: Enable shadow stack support

hjl-tools opened the pull request:
Intel Control-flow Enforcement Technology (CET):

https://software.intel.com/en-us/articles/intel-sdm

contains shadow stack (SHSTK).  When SHSTK is enabled, return address on
normal stack must match the one on shadow stack.  When SHSTK is enabled
at the compile time, this patch updates sljit with:

1. Before JIT return, if shadow stack is enabled, adjust shadow stack
so that the top of show stack matches the top of normal stack.  NB: It
is safe since shadow stack is read-only, you can only skip stack frame,
not jump to anywhere else.
2. Fix X86_32 emit_x86_instruction to properly encode (%ebp) so that
we can use TMP_REG1 as scratch register to unwind show stack in
sljit_emit_return.  NB: ECX is unusable since it isn't a scratch
register in sljit as shown in sljit test:

$ ./bin/sljit_test
Pass -v to enable verbose, -s to disable this hint.

test51 case 2 failed
SLJIT tests: 1 (2%) tests are FAILED on x86 32bit (little endian + unaligned) ABI:fastcall (with fpu)

Tested with

$ CC="gcc -Wl,-z,cet-report=error -fcf-protection -mshstk" make

on x86-32 and x86-64 Linux CET machines.

test67 isn't compatible with CET since it uses sljit_emit_fast_return to
return from JIT.  We can't adust shadow stack in sljit_emit_fast_return
since it modifies flags, but sljit_emit_fast_return must preserve flags
as checked by test62.

zherczeg commented:
Similar to #13, I would prefer this feature as user controlled. As for sljit_emit_return, we could use a flag which tells that the shadow stack needs to be cleaned. As for fast calls, we would need another function, which prepares the shadow stack before return, but does not return, so the user still can do some tasks (e.g. setting status flag). If I understand correctly, you can update the shadow stack anytime, the only thing matters is that when the return is executed, the shadow stack must know the location.

If you are ok with this change, I might delete the fast return function and create a new one where you can pass an opcode and has a source. I might move the prefetch instruction here as well later.

hjl-tools commented:
> Similar to #13, I would prefer this feature as user controlled. As for sljit_emit_return, we could use a flag which tells that the shadow stack needs to be cleaned. As for fast calls, we would need another function, which prepares the shadow stack before return, but does not return, so the user still can do some tasks (e.g. setting status flag). If I understand correctly, you can update the shadow stack anytime, the only thing matters is that when the return is executed, the shadow stack must know the location.

Adding a flag to sljit_emit_return should work.  A new function to prepare for fast return can
make shadow stack work.

I can only update shadow stack anytime if the return address is known or at a known location
on normal stack.  Since return address is passed to fast return, it isn't a problem.

> If you are ok with this change, I might delete the fast return function and create a new one where you can pass an opcode and has a source. I might move the prefetch instruction here as well later.

Yes, please.  Thanks.

zherczeg commented:
A patch is landed which introduces `check_sljit_emit_op_src`. You can add a new opcode here which updates the shadow stack according to the src value passed to this function. Would it be possible to add another `sljit_emit_op0` opcode which updates the shadow stack before `sljit_emit_return`? This way we would not need to add a flag (and we could do things after the shadow stack is updated but before we return).

Another question: do you use the ebp adressing on 32 bit frequently? Because if you don't, we could just emit the instruction directly, so there is no need to change the generic instruction emitter. I worry a bit about the FIXME comment.

hjl-tools commented:
> A patch is landed which introduces `check_sljit_emit_op_src`. You can add a new opcode here which updates the shadow stack according to the src value passed to this function. Would it be possible to add another `sljit_emit_op0` opcode which updates the shadow stack before `sljit_emit_return`? This way we would not need to add a flag (and we could do things after the shadow stack is updated but before we return).

I don't think `check_sljit_emit_op_src` help here since it can't tell me if shadow stack needs
adjustment.    `sljit_emit_op0` can work.  One can add it before return to adjust shadow
stack.

> Another question: do you use the ebp adressing on 32 bit frequently? Because if you don't, we could just emit the instruction directly, so there is no need to change the generic instruction emitter. I worry a bit about the FIXME comment.

I don't think (%ebp) will be used in other places and I can encode (%ebp) directly.

BTW, should it be an error when (%ebp/r13) is used in the generic instruction emitter?
It took me a while to figure what went wrong.

zherczeg commented:
> I don't think `check_sljit_emit_op_src` help here since it can't tell me if shadow stack needs
> adjustment. `sljit_emit_op0` can work. One can add it before return to adjust shadow
> stack.

As for fast return, the return address is not stored on the stack. It is provided in a register or memory location. This fits better for CPUs with a link register. I suspect you need to know the return address, hence I recommended to add a new opcode to `check_sljit_emit_op_src`. As for normal return, the return address is on the stack, you just need to calculate its position.

> BTW, should it be an error when (%ebp/r13) is used in the generic instruction emitter?
> It took me a while to figure what went wrong.

Yes, an SLJIT_ASSERT could be added here. As far as I know the r13 is not an issue, because the REX prefix enough to determine that a register >= 8 is used, and you don't need any special casing.

hjl-tools commented:
> > I don't think `check_sljit_emit_op_src` help here since it can't tell me if shadow stack needs
> > adjustment. `sljit_emit_op0` can work. One can add it before return to adjust shadow
> > stack.
> 
> As for fast return, the return address is not stored on the stack. It is provided in a register or memory location. This fits better for CPUs with a link register. I suspect you need to know the return address, hence I recommended to add a new opcode to `check_sljit_emit_op_src`. As for normal return, the return address is on the stack, you just need to calculate its position.

I added SLJIT_FAST_RETURN_SKIP_FRAMES and use it in in test67 to skip stack frames
for fast return.

> > BTW, should it be an error when (%ebp/r13) is used in the generic instruction emitter?
> > It took me a while to figure what went wrong.
> 
> Yes, an SLJIT_ASSERT could be added here. As far as I know the r13 is not an issue, because the REX prefix enough to determine that a register >= 8 is used, and you don't need any special casing.

I added SLJIT_ASSERT for unsupported (%esp) and hand coded "mov 0x0(%ebp),%ebp".
Everything works now.

zherczeg commented on sljit_src/sljitNativeX86_32.c:
```diff
@@ -310,26 +310,39 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_return(struct sljit_compiler *comp
-	if (compiler->saveds > 1 || compiler->scratches > 10)
-		POP_REG(reg_map[SLJIT_S1]);
-	if (compiler->saveds > 2 || compiler->scratches > 9)
-		POP_REG(reg_map[SLJIT_S2]);
```
Do we need these changes? I thought the assertion below is enough.

hjl-tools replied:
We have

```
POP some registers    <-- Separate instruction buffer
Adjust shadow stack if needed <-- Separate instruction buffer
POP TMP_REG1
RETURN
```

When allocating the instruction buffer for `POP some registers`, the buffer size must be
precise.  There must be no unused space in  `POP some registers` instruction buffer.
The bug is latent since a single instruction buffer used to covere:

```
POP some registers    <-- Separate instruction buffer
POP TMP_REG1
RETURN
```

with some unused space after RETURN.

zherczeg replied:
Ah, I am sorry. I thought this code is related to emit_x86_instruction. So we make `adjust_shadow_stack` mandatory for `sljit_emit_return`? Could we make it an op0 operation, which can be run by the user when it is necessary?

hjl-tools replied:
adjust_shadow_stack must know exactly where return address is on stack.  How can a user
tell and pass the info to the op0 operation?

zherczeg replied:
The op0 has no operand. The point is, the user pass no info, but you can compute everything you need. See sljit_set_context:

https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitNativeX86_32.c#L247

`compiler->local_size` contains the stack frame size except the saved registers and the return address. The following code shows how the stack is adjusted before return:

https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitNativeX86_32.c#L302

So you can load the return address by adding a constant to esp:
mov ebp, [esp + compiler->local_size + saveds_size]

Note: compiler->local_size + saveds_size is a compile time constant.

saveds_size for x86-32:
https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitNativeX86_32.c#L278

When SLJIT_F64_ALIGNMENT flag is set, it is a bit more complicated because you need two loads:
mov ebp, [esp + compiler->local_size]
mov ebp, [ebp + saveds_size]

As for 64 bit, it is easier because you need only one load all the time (formulas are different though).

zherczeg replied:
`saveds = (2 + (scratches > 9 ? (scratches - 9) : 0) + (saveds <= 3 ? saveds : 3)) * sizeof(sljit_uw);`

The 2 in this formula represents the return adress and ebp, so you might need to add only 1:
`saveds = (1 + (scratches > 9 ? (scratches - 9) : 0) + (saveds <= 3 ? saveds : 3)) * sizeof(sljit_uw);`

But please check, I just give the general idea, haven't checked it yet.

hjl-tools replied:
Can you add a testcase with sljit_emit_return with frames skipped?  The current
testcases don't cover this case.

zherczeg replied:
I have added a new test.

hjl-tools replied:
```
static void test68(void)
{
        /* Test skipping returns from fast calls (return type is normal). */
        executable_code code;
        struct sljit_compiler *compiler;
        struct sljit_jump *call, *jump;
        struct sljit_label *label;
        int i;

        if (verbose)
                printf("Run test67\n");
                                   ^^^^^ Should be test68.
````

zherczeg replied:
Thanks. I will fix it later.

hjl-tools replied:
I added SLJIT_SKIP_FRAMES_BEFORE_RETURN.   It passed testes.

zherczeg commented on sljit_src/sljitNativeX86_common.c:
```diff
@@ -2220,6 +2363,8 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_op_src(struct sljit_compiler *comp
 
 	switch (op) {
+	case SLJIT_FAST_RETURN_SKIP_FRAMES:
+		FAIL_IF(adjust_shadow_stack(compiler, src, srcw, 0));
```
Maybe I wasn't clear, but I want this as a separate operation, and not part of `SLJIT_FAST_RETURN`.

Ideally you could do the following:
sljit_emit_op_src(compiler, SLJIT_SKIP_FRAMES_BEFORE_FAST_RETURN, src, srcw);
/* Do other things, e.g. set flags, since SLJIT_SKIP_FRAMES_BEFORE_FAST_RETURN may destroy them. */
sljit_emit_op_src(compiler, SLJIT_FAST_RETURN, src, srcw);

This provides quite a flexibility since users can do the shadow stack update anytime, not when the return is actually performed.

hjl-tools replied:
Fixed.

zherczeg commented on sljit_src/sljitNativeX86_32.c:
```diff
@@ -370,6 +370,8 @@ static sljit_u8* emit_x86_instruction(struct sljit_compiler *compiler, sljit_s32
 		&& (flags & (EX86_PREF_F2 | EX86_PREF_66)) != (EX86_PREF_F2 | EX86_PREF_66)
 		&& (flags & (EX86_PREF_F3 | EX86_PREF_66)) != (EX86_PREF_F3 | EX86_PREF_66));
+	/* We don't support (%esp). */
+	SLJIT_ASSERT(!(b & SLJIT_MEM) || immb || reg_map[b & REG_MASK] != 5);
```
Isn't this in another patch?

hjl-tools replied:
Fixed.

zherczeg commented on sljit_src/sljitNativeX86_32.c:
```diff
@@ -893,3 +895,36 @@ static sljit_s32 emit_fast_return(struct sljit_compiler *compiler, sljit_s32 src
+		 + (compiler->saveds <= 3 ? compiler->saveds : 3)) * 4 + 4;
+
+	/* Adjust shadow stack if needed.  */
+	return adjust_shadow_stack(compiler, SLJIT_UNUSED, 0, base, size);;
```
Single semicolon is enough.

hjl-tools replied:
Fixed.

zherczeg commented on sljit_src/sljitNativeX86_64.c:
```diff
@@ -897,3 +897,19 @@ static sljit_s32 emit_mov_int(struct sljit_compiler *compiler, sljit_s32 sign,
+	size = compiler->local_size;
+	tmp = compiler->scratches;
+	if (tmp >= SLJIT_FIRST_SAVED_REG)
+		size += (tmp - SLJIT_FIRST_SAVED_REG + 1) * 8;
```
Use `sizeof(sljit_uw)` instead of 8.

hjl-tools replied:
Fixed.

zherczeg commented on sljit_src/sljitNativeX86_64.c:
```diff
@@ -897,3 +897,19 @@ static sljit_s32 emit_mov_int(struct sljit_compiler *compiler, sljit_s32 sign,
+		size += (tmp - SLJIT_FIRST_SAVED_REG + 1) * 8;
+	tmp = compiler->saveds < SLJIT_NUMBER_OF_SAVED_REGISTERS ? (SLJIT_S0 + 1 - compiler->saveds) : SLJIT_FIRST_SAVED_REG;
+	if (SLJIT_S0 >= tmp)
+		size += (SLJIT_S0 - tmp + 1) * 8;
```
Ditto.

hjl-tools replied:
Fixed,

zherczeg commented on sljit_src/sljitLir.c:
```diff
@@ -1161,9 +1161,10 @@ static SLJIT_INLINE CHECK_RETURN_TYPE check_sljit_emit_op0(struct sljit_compiler
 	CHECK_ARGUMENT(GET_OPCODE(op) < SLJIT_LMUL_UW || GET_OPCODE(op) >= SLJIT_ENDBR || compiler->scratches >= 2);
-	if (GET_OPCODE(op) >= SLJIT_LMUL_UW && GET_OPCODE(op) <= SLJIT_DIV_SW)
+	if ((GET_OPCODE(op) >= SLJIT_LMUL_UW && GET_OPCODE(op) <= SLJIT_DIV_SW)
+	    || op == SLJIT_SKIP_FRAMES_BEFORE_RETURN)
```
Please add the name to `op0_names`

hjl-tools replied:
Fixed.

zherczeg commented on sljit_src/sljitLir.c:
```diff
@@ -1317,10 +1318,12 @@ static SLJIT_INLINE CHECK_RETURN_TYPE check_sljit_emit_op_src(struct sljit_compi
 
-	if (op == SLJIT_FAST_RETURN)
+	if (op == SLJIT_FAST_RETURN
+	    || op == SLJIT_SKIP_FRAMES_BEFORE_FAST_RETURN)
```
Please add the name to `op_src_names`

hjl-tools replied:
Fixed.

zherczeg commented:
I have added some updates, please check them.

One more thing came to my mind. We silently overwrite ecx for x86-32 in some cases. Could we push/pop this register?

zherczeg commented:
`[esp+0]` is also reserved for temporaries, we could also use this memory location.

hjl-tools commented:
> `[esp+0]` is also reserved for temporaries, we could also use this memory location.

See https://github.com/zherczeg/sljit/pull/19
